code/game/objects/structures/grille.dm

```diff
@@ -10,208 +10,215 @@
+					  "<span class='warning'>You smash against \the [src].</span>", \
 					  "You hear twisting metal.")
-
-	health -= M.melee_damage_upper
-	healthcheck()
+	health -= rand(M.melee_damage_lower, M.melee_damage_upper)
+	healthcheck(hitsound = 1)
+	shock(M, 100)
 	return
 
 
-/obj/structure/grille/CanPass(atom/movable/mover, turf/target, height=1.5, air_group = 0)
-	if(air_group || (height==0)) return 1
+/obj/structure/grille/CanPass(atom/movable/mover, turf/target, height = 1.5, air_group = 0)
+	if(air_group || (height == 0))
+		return 1
 	if(istype(mover) && mover.checkpass(PASSGRILLE))
 		return 1
 	else
 		if(istype(mover, /obj/item/projectile))
-			return prob(30)
+			return prob(66) //Fairly hit chance
 		else
 			return !density
 
 /obj/structure/grille/bullet_act(var/obj/item/projectile/Proj)
-	if(!Proj)	return
-	src.health -= Proj.damage*0.2
-	healthcheck()
+	if(!Proj)
+		return
+	health -= Proj.damage //Just use the projectile damage, it already has high odds of "missing"
+	healthcheck(hitsound = 1)
 	return 0
 
 /obj/structure/grille/attackby(obj/item/weapon/W as obj, mob/user as mob)
 	user.delayNextAttack(8)
 	if(iswirecutter(W))
-		if(!shock(user, 100))
+		if(!shock(user, 100)) //Prevent user from doing it if he gets shocked
 			playsound(loc, 'sound/items/Wirecutter.ogg', 100, 1)
-			new /obj/item/stack/rods(loc, 2)
+			getFromPool(/obj/item/stack/rods, get_turf(src), broken ? 1 : 2) //Drop the rods, taking account on whenever the grille is broken or not !
 			returnToPool(src)
+			return
+		return //Return in case the user starts cutting and gets shocked, so that it doesn't continue downwards !
 	else if((isscrewdriver(W)) && (istype(loc, /turf/simulated) || anchored))
 		if(!shock(user, 90))
 			playsound(loc, 'sound/items/Screwdriver.ogg', 100, 1)
 			anchored = !anchored
-			user.visible_message("<span class='notice'>[user] [anchored ? "fastens" : "unfastens"] the grille.</span>", \
-								 "<span class='notice'>You have [anchored ? "fastened the grille to" : "unfastened the grill from"] the floor.</span>")
+			user.visible_message("<span class='notice'>[user] [anchored ? "fastens" : "unfastens"] the grille [anchored ? "to" : "from"] the floor.</span>", \
+			"<span class='notice'>You [anchored ? "fasten" : "unfasten"] the grille [anchored ? "to" : "from"] the floor.</span>")
 			return
 
-//window placing begin
-	else if(istype(W,/obj/item/stack/sheet/glass))
-		var/dir_to_set = 1
+//Window placement
+	else if(istype(W, /obj/item/stack/sheet/glass))
+		var/dir_to_set
 		if(loc == user.loc)
-			dir_to_set = user.dir
+			dir_to_set = user.dir //Whatever the user is doing, return the "normal" window placement output
 		else
-			if( ( x == user.x ) || (y == user.y) ) //Only supposed to work for cardinal directions.
-				if( x == user.x )
-					if( y > user.y )
-						dir_to_set = 2
+			if((x == user.x) || (y == user.y)) //Only supposed to work for cardinal directions, aka can't lay windows in diagonal directions

```

Comment:
`if(get_dir(user, src) in cardinal)`
`    new window whatever`
`    window.dir = get_dir(user, src)`